data/state/accountsDB.go

```diff
@@ -379,13 +369,39 @@ func (adb *AccountsDB) RemoveAccount(address []byte) error {
 	}
 	adb.journalize(entry)
 
+	err = adb.removeCode(acnt)
+	if err != nil {
+		return err
+	}
+
 	log.Trace("accountsDB.RemoveAccount",
 		"address", hex.EncodeToString(address),
 	)
 
 	return adb.mainTrie.Update(address, make([]byte, 0))
 }
 
+func (adb *AccountsDB) removeCode(acnt AccountHandler) error {
+	baseAcc, ok := acnt.(baseAccountHandler)
+	if !ok {
+		return nil
+	}
+
+	oldCodeHash := baseAcc.GetCodeHash()
+	unmodifiedOldCodeEntry, err := adb.updateOldCodeEntry(oldCodeHash)
+	if err != nil {
+		return err
+	}
+
+	codeChangeEntry, err := NewJournalEntryCode(unmodifiedOldCodeEntry, oldCodeHash, nil, adb.mainTrie, adb.marshalizer)
```

Comment:
Should the newJournalEntryCode have the newCodeHash=nil every time? It does not have necessarily NumReferences<=0 and it is not specific for the current AccountHandler, right?

Reply:
oldCode refers to the code that the account had. After an account is removed, the num references to that code must be decreased, and if it is 0, removed. newCode refers to the code that is added to the account. Since the removed account will not refer to any code, newCodeHash is nil.

Reply:
ok